openedx/core/djangoapps/content_libraries/api.py

```diff
@@ -1062,6 +1067,69 @@ def revert_changes(library_key):
     )
 
 
+def update_collection_contents(
+    library: ContentLibraryMetadata,
+    collection_pk: int,
+    usage_keys: list[UsageKeyV2],
+    remove=False,
+) -> int:
+    """
+    Associates the Collection with Components for the given UsageKeys.
+
+    By default the Components are added to the Collection.
+    If remove=True, the Components are removed from the Collection.
+
+    Raises:
+    * Collection.DoesNotExist if no Collection with the given pk is found in the given library.
+    * Component.DoesNotExist if any of the given usage_keys don't correspond to a Component in the
+      library/learning_package.
+    """
+    learning_package = library.learning_package
+    collections_qset = authoring_api.get_learning_package_collections(
+        learning_package.id,
+    ).filter(pk=collection_pk)
+
+    collection = collections_qset.first()
+    if not collection:
+        raise Collection.DoesNotExist()
+
+    # Fetch the Component.key values for the provided UsageKeys.
+    component_keys = []
+    for usage_key in usage_keys:
+        # Parse the block_family from the key to use as namespace.
+        block_type = BlockTypeKey.from_string(str(usage_key))
+
+        # Can raise Component.DoesNotExist
+        component = authoring_api.get_component_by_key(
+            learning_package.id,
+            namespace=block_type.block_family,
+            type_name=usage_key.block_type,
+            local_key=usage_key.block_id,
+        )
+        component_keys.append(component.key)
+
+    # Note: Component.key matches its PublishableEntity.key
+    contents_qset = learning_package.publishable_entities.filter(
+        key__in=component_keys,
+    )
+    if not contents_qset.count():
+        raise Component.DoesNotExist()
+
+    if remove:
+        count = authoring_api.remove_from_collections(collections_qset, contents_qset)
+    else:
+        count = authoring_api.add_to_collections(collections_qset, contents_qset)
+
+    # Emit a CONTENT_OBJECT_TAGS_CHANGED event for each of the objects added/removed
+    object_keys = contents_qset.values_list("key", flat=True)
+    for object_key in object_keys:
+        CONTENT_OBJECT_TAGS_CHANGED.send_event(
+            content_object=ContentObjectData(object_id=object_key),
+        )
+
```

Comment:
I don't think I get why we are dispatching `CONTENT_OBJECT_TAGS_CHANGED` here. Maybe you mean `LIBRARY_COLLECTION_UPDATED`?

There is a discussion about this event [here](https://github.com/openedx/openedx-events/pull/386#issuecomment-2315828221).

Reply:
@rpenido Yep, that's the current behaviour, but we're planning to update it with https://github.com/openedx/modular-learning/issues/230. I've made a start at that here: https://github.com/open-craft/edx-platform/compare/jill/collection-components-rest...open-craft:edx-platform:jill/collection-components-search

Reply:
@rpenido I will rename `CONTENT_OBJECT_TAGS_CHANGED` to `CONTENT_OBJECT_ASSOCIATIONS_CHANGED`, and add a field indicating whether it was the `tag` or `collection` that changed when I do https://github.com/openedx/modular-learning/issues/230.